File: Proyecto/Sprint.5/sos/src/classes/models/Board.ts
import { Letter } from "@/classes/enums/Letter";

/**
 * @class Board
 * @classdesc Representa un tablero de juego.
 */
export class Board {
	private readonly rows: number;
	private readonly columns: number;
	private readonly grid: Letter[][];

	/**
	 * Crear un tablero de juego.
     * 
	 * @constructor
	 * @param {number} rows - Número de filas del tablero (opcional).
	 * @param {number} columns - Número de columnas del tablero (opcional).
	 */
	constructor(rows: number = 3, columns: number = 3) {
		this.rows = rows;
		this.columns = columns;
		this.grid = new Array(this.rows);
		this.grid = new Array(rows)
			.fill(null)
			.map(() => new Array(columns).fill(Letter.EMPTY));
	}

	/**
	 * Retorna el número de filas del tablero.
     * 
	 * @returns {number} Número de filas del tablero.
	 */
	public getRows(): number {
		return this.rows;
	}

	/**
	 * Retorna el número de columnas del tablero.
     * 
	 * @returns {number} Número de columnas del tablero.
	 */
	public getColumns(): number {
		return this.columns;
	}

	/**
	 * Retorna el tablero de juego.
     * 
	 * @returns {Letter[][]} Tablero de juego.
	 */
	public getGrid(): Letter[][] {
		return this.grid;
	}

    /**
     * Retorna `true` si el tablero está vacío.
     * 
     * @returns {boolean} `true` si el tablero está vacío, `false` en caso contrario.
     */
    public isEmpty(): boolean {
        return this.grid.every((row) => row.every((cell) => cell === ""));
    }

    /**
     * Retorna `true` si el tablero está lleno.
     * 
     * @returns {boolean} `true` si el tablero está lleno, `false` en caso contrario.
     */
    public isFull(): boolean {
        return this.grid.every((row) => row.every((cell) => cell !== ""));
    }

	/**
	 * Establece el valor de una celda del tablero.
     * 
	 * @param {number} row Posición de la fila.
	 * @param {number} column Posición de la columna.
	 * @param {Letter} value Valor a asignar.
	 */
	public setCell(row: number, column: number, value: Letter): void {
		this.grid[row][column] = value;
	}

	/**
	 * Obtiene el valor de una celda del tablero.
     * 
	 * @param {number} row Posición de la fila.
	 * @param {number} column Posición de la columna.
	 * @returns {string} Valor de la celda.
	 */
	public getCell(row: number, column: number): string {
		return this.grid[row][column];
	}

}

File: Actividades/Semana 8/8.1 ExamenParcial/TicTacToe/src/main/java/com/jhaner/Board.java
package com.jhaner;

/**
 * Representa un tablero de juego con celdas que pueden ser ocupadas por diferentes valores.
 * El tablero tiene un número de filas y columnas determinado al inicio del juego.
 */
public class Board {

    private final int rows;
    private final int cols;
    private final Letter[][] grid;

    /**
     * {@code @constructor}
     * Crear un tablero con el numero de filas y columans asignadas.
     *
     * @param rows el numero de filas del tablero.
     * @param cols el numero de columnas del tablero.
     */
    public Board(int rows, int cols) {
        this.rows = rows;
        this.cols = cols;
        this.grid = new Letter[rows][cols];
    }

    /**
     * Devuelve las filas del tablero.
     *
     * @return en numero de filas.
     */
    public int getRows() {
        return rows;
    }

    /**
     * Devuelve las columnas del tablero.
     *
     * @return el numero de columnas.
     */
    public int getCols() {
        return cols;
    }

    /**
     * Escribe en una casilla especifica del tablero.
     *
     * @param row   la fila de la celda.
     * @param col   la columna de la celda.
     * @param value el valor a escribir en la celda.
     */
    public void setCell(int row, int col, Letter value){
        this.grid[row][col] = value;
    }

    /**
     * Devuelve una casilla especifica del tablero
     *
     * @param row la fila de la celda.
     * @param col la columna de la celda.
     * @return el valor de la celda correspondiente a las coordenadas dadas.
     */
    public Letter getCell(int row, int col){
        return grid[row][col];
    }

    /**
     * Devuelve true si la celda esta vacia y false en caso contrario.
     *
     * @param row la fila de la celda.
     * @param col la columna de la celda.
     * @return true si la celda esta vacia, false en caso contrario.
     */
    public boolean isEmptyCell(int row, int col){
        return this.getCell(row, col) == Letter.EMPTY;
    }

}
